drivers/sensor/ti_hdc20xx/ti_hdc20xx.c

```diff
@@ -33,15 +44,28 @@ LOG_MODULE_REGISTER(TI_HDC20XX, CONFIG_SENSOR_LOG_LEVEL);
 #define TI_HDC20XX_TEMP_SCALE		165U
 
 struct ti_hdc20xx_config {
-	const struct device *bus;
-	uint16_t i2c_addr;
+	struct i2c_dt_spec bus;
+	struct gpio_dt_spec gpio_int;
 };
 
 struct ti_hdc20xx_data {
-	int32_t t_sample;
-	int32_t rh_sample;
+	struct gpio_callback cb_int;
+	struct k_sem sem_int;
+
+	uint16_t t_sample;
+	uint16_t rh_sample;
 };
 
+static void ti_hdc20xx_int_callback(const struct device *dev,
+				    struct gpio_callback *cb, uint32_t pins)
+{
+	struct ti_hdc20xx_data *data = CONTAINER_OF(cb, struct ti_hdc20xx_data, cb_int);
+
+	ARG_UNUSED(pins);
```

Comment:
I think this callback should be used to call `ti_hdc20xx_sample_fetch(dev, SENSOR_CHAN_ALL)`, or at least submit a `struct k_work` to do so in the system workqueue.

Reply:
The driver doesn't work (yet) in trigger mode where the chip is running measurements itself at regular intervals. The measurement is actually triggered by `ti_hdc20xx_sample_fetch`, so the interrupt pin is only used to wait for the conversion to finish. In that regards, it doesn't make sense to call this function again, the semaphore is there to signal the data arrived.

Overall this shows an issue in the sensor API, as it tries to support both measurements scheduled by the MCU and by the sensor. For some sensors only one of the two solutions is possible (depending if they have an interrupt line and it if it wired), for some other both are possible. And we ended up we the two strategies implemented depending on the driver.

In practice there is a usage for both. One might want to let the sensor schedule the measurements to lower the MCU work. One might want to schedule the measurements from the host to have more flexibility in the timing, or two synchronize the sampling of different sensors.

Reply:
> The driver doesn't work (yet) in trigger mode where the chip is running measurements itself at regular intervals. The measurement is actually triggered by `ti_hdc20xx_sample_fetch`, so the interrupt pin is only used to wait for the conversion to finish. In that regards, it doesn't make sense to call this function again, the semaphore is there to signal the data arrived.

I see - yeah. There is also a `sensor_trigger_set_t` in the sensor API that is used to set the trigger, which could either be a bit of `k_work_delayable` or kind of a free-running trigger if the sensor itself supports automatically re-arming itself. In the latter case, it makes sense to use interrupts, but it might not necessarily in the former case. If PM is configured, then polling mode would need to be switched off, etc.

> Overall this shows an issue in the sensor API, as it tries to support both measurements scheduled by the MCU and by the sensor. For some sensors only one of the two solutions is possible (depending if they have an interrupt line and it if it wired), for some other both are possible. And we ended up we the two strategies implemented depending on the driver.

Other drivers support both, but it may need a bit of work in the long run. For now should be ok though.

Thanks for clarifying.

Reply:
> Other drivers support both, but it may need a bit of work in the long run. For now should be ok though.

Do you have a pointer to a good implementation? The drivers I have found select the running mode through a Kconfig option, but that just scale for drivers supporting multiple instances.

An idea I got is to use the SENSOR_ATTR_SAMPLING_FREQUENCY. It could already be used to switch between the various the sensor scheduled frequencies and the special value 0 can be used to mean MCU scheduled measurements. Would that make sense?

The other alternative would be to add a new `sensor_sample_measure` function to the API, changing `sensor_sample_fetch` to mean only fetch. That way when the measurements are scheduled by the MCU, `sensor_sample_measure` would need to be called before `sensor_sample_fetch`, while when the measurements are scheduled by the sensor, this would not be needed.